BDArmory/Control/BDModulePilotAI.cs

```diff
@@ -2328,7 +2333,11 @@ void FlyToTargetVessel(FlightCtrlState s, Vessel v)
                         float magnifier = Mathf.Clamp(targetAngVel, 1f, 2f);
                         magnifier += ((magnifier - 1f) * Mathf.Sin(Time.time * 0.75f));
                         if (BDArmorySettings.DEBUG_TELEMETRY || BDArmorySettings.DEBUG_AI) debugString.AppendLine($"targetAngVel: {targetAngVel:F4}, magnifier: {magnifier:F2}");
+                        target = Quaternion.FromToRotation(weapon.fireTransforms[0].forward, vesselTransform.up) * (target - vesselTransform.position) + vesselTransform.position;
```

Comment:
This should come after target is adjusted by `leadOffset` on the next line below as it's adjusting the target for the weapon's offset angle after bullet drop is accounted for. Also the `angleToTarget = ...` line should be using `weapon.fireTransforms[0]` instead of `vesselTransform`.

However, in testing, the AI doesn't properly handle the adjusted direction decently unless the angle is smallish (<20° or so) and positive as the rolling to adjust to the requested direction changes that direction, giving a feedback loop that gives oscillations for weapons pitching upwards too much and spirals for weapons pitching downwards. (This happens for me both when using rotations or the commented out vector additions.)
I think the way to fix this is that the rotation of the weapon relative to the craft needs to be a local rotation relative to a `LookRotation` towards the target with "up" being the target projected onto the `vesselTransform.up` plane. (I know that's not a particularly good description, I'm still trying to visualise it and I'm tired... I'll try to figure it out tomorrow.)

Reply:
Ah, good catch; didn't get fixed during one of the reversions.
Haven't noticed the roll much when the weapon is pointing up (haven't tested a weapon pointing down) though a weapon pointing to the side did tend to have the test craft perma-barrelrolling unless the target was flying straight and level. There's probably an element of craft design in this; a stock Aeris 3A definitely had problems, a custom craft with centered wings and symmetrical rudders fared much better.
Part of the roll is probably coming from roll controls in FlyToPosition happening after the adjusted target is handed off and the AI seeking to avoid negative pitch(for ventrally pointing weapons) and seeking to keep ventral aspect of craft pointed towards target when maneuvering (for dorsal guns).

Reply:
So, I was thinking about this last night and I think the issue isn't with the code, but with the physics of the situation.
Basically, for fixed guns that aren't aligned with the vessel's forward axis, the plane has to try to aim for and maintain a fly-to direction that is offset from the target.
- For small upward pitched guns, this means that the fly-to point pretty much lags the target position by a small amount and the way that the plane adjusts its direction is somewhat stable depending on the airframe.
- For larger upwardly pitched guns, the amount of roll the plane does to adjust its direction induces an oscillation by pushing the fly-to point too far in the opposite direction.
- For downward pitched guns, the fly-to point leads the target and the way the plane adjusts its direction pushes the fly-to point further away in a positive feedback loop causing it to spiral.
- For guns offset in yaw, something inbetween the above cases happens.

Essentially, depending on the airframe, I think there's a smallish region of stability in the pitch-up direction (with some smallish yaw component) where the ability of the airframe to correct a deviation away from the required fly-to point is within the region where the airframe can stably make adjustments.

Reply:
It would be very much airframe dependent, yes. The initial intention of the changes was solely to correct for small offsets - gun is pointing upwards at ~20deg, which should be well within the ability to aim at a target (just leading a bit farther back), or has wingtip-mounted guns where the gun is laterally offset from the longitudinal axis by 6m or something, weapon would get pointed in the right direction, but wouldn't account for the lateral offset so the weapon would need a larger firingAngle/wouldn't fire. 
That it also worked for very silly things like mounting the weapon pointing straight up or broadside was more happy accident than deliberate intent, and if someone does go for the silly option, then it's on them do do the necessary engineering to facilitate it.